x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTrunc.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.sql.expression.function.scalar.datetime;
+
+import org.elasticsearch.xpack.sql.expression.Expression;
+import org.elasticsearch.xpack.sql.expression.Expressions;
+import org.elasticsearch.xpack.sql.expression.Nullability;
+import org.elasticsearch.xpack.sql.expression.function.scalar.BinaryScalarFunction;
+import org.elasticsearch.xpack.sql.expression.gen.pipeline.Pipe;
+import org.elasticsearch.xpack.sql.expression.gen.script.ScriptTemplate;
+import org.elasticsearch.xpack.sql.tree.NodeInfo;
+import org.elasticsearch.xpack.sql.tree.Source;
+import org.elasticsearch.xpack.sql.type.DataType;
+
+import java.time.ZoneId;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.Set;
+
+import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
+import static org.elasticsearch.xpack.sql.expression.TypeResolutions.isDate;
+import static org.elasticsearch.xpack.sql.expression.TypeResolutions.isString;
+import static org.elasticsearch.xpack.sql.expression.function.scalar.datetime.DateTruncProcessor.process;
+import static org.elasticsearch.xpack.sql.expression.gen.script.ParamsBuilder.paramsBuilder;
+
+public class DateTrunc extends BinaryScalarFunction {
+
+    public enum DatePart {
+
+        MILLENNIUM("millennia"),
+        CENTURY("centuries"),
+        DECADE("decades"),
+        YEAR("years", "yy", "yyyy"),
+        QUARTER("quarters", "qq", "q"),
+        MONTH("months", "mm", "m"),
+        WEEK("weeks", "wk", "ww"),
+        DAY("days", "dd", "d"),
+        HOUR("hours", "hh"),
+        MINUTE("minutes", "mi", "n"),
+        SECOND("seconds", "ss", "s"),
+        MILLISECOND("milliseconds", "ms"),
+        MICROSECOND("microseconds", "mcs"),
+        NANOSECOND("nanoseconds", "ns");
+
+        private Set<String> aliases;
+
+        DatePart(String... aliases) {
+            this.aliases = Set.of(aliases);
+        }
+
+        public Set<String> aliases() {
+            return aliases;
+        }
+
+        public static DatePart resolveTruncate(String truncateTo) {
```

Comment:
An `EnumMap` is better than iterating and converting the string to lowercase every time.

Reply:
Hm, I don't get how the EnumMap can be beneficial here. We can receive any string like: `w` or `milliseconds` or `MILLENNIUM` and we need to find out if it resolves to an Enum name or its aliases. I was thinking of creating a `HashMap<String, Enum>` for faster resolution to avoid the iteration, but we still need to do the `lowerCase()`.

Can you please explain more your suggested solution? (probably I'm missing something)

Reply:
Right, that is what I meant, an map of enums: `Map<String, Enum>` (not `EnumMap since the keys are not `enum`s themselves).
The point of lowercase is you don't have to do equalsIgnoreCase or lowercase the enum name since you can do that directly when initializing the map - this is the lowercasing I meant not that of the argument.